### PR TITLE
Added tests for Static.

### PR DESCRIPTION
**Pull request by chriselrod**
@Tokazama 
This is the code for `Static`. The `OptionallyStaticUnitRange` could probably be cleaned up further.

Currently `known_*` functions still return `nothing` or values, but I added `static_*` that use these and return either `Static` numbers (if `known_` returns a value) or values (if they return `nothing`). These will likely often be more convenient to use, and seem strictly more powerful.

### CODE REVIEW

**Comment by Tokazama** [comment]
This is great! The `static_*` methods seem like a nice way to ensure that static numbers are only used when asking for them.

Do you think it would be useful to support other integer types? If so I'm thinking of something that's still limited to types defined in base like one of the following

```julia
struct Static{N} <: Integer
    function Static{N}()  where {N}
        @assert isprimitivetype(typeof(N)) "$N is not a primitive type"
        return new{N}()
    end
end
```

```julia
struct Static{N} <: Integer
    function Static{N}()  where {N}
        @assert isa(N, Base.BitIntegerType) "$N is not a primitive type"
        return new{N}()
    end
end
```

**Comment by chriselrod** [comment]
The reason I restricted it to `Int` was because of defining all the methods like `Base.:(::Static{0}, x) = Static{0}()`.
I also realize now I should also add more generic methods (i.e., where `x` is untyped like in that example). 

I also have a lot of trust that dispatch will be type stable.
But branches should be fine, too. If we go for allowing more integer types, we can use branch definitions such as:
```julia
@inline function Base.:(x::Static{M}, y) where {M}
    if iszero(M)
        x
    elseif isone(M)
        y
    else
        M * y
    end
end
```
and I think it's unlikely inference will fail.

**Comment by timholy** [comment]
Allowing other Integer subtypes risks increasing your compile-time latency and likelihood of difficult-to-track-down bugs; try debugging why
```
julia> Val(3)
Val{3}()

julia> Val(Int16(3))
Val{3}()

julia> typeof(Val(3))
Val{3}

julia> typeof(Val(Int16(3)))
Val{3}
```
aren't the same type. Unless there's a compelling reason, diversifying the Integer types seems like a lot of pain for little or no benefit.

**Comment by chriselrod** [comment]
I'll revert adding support for more integer types.

**Comment by Tokazama** [comment]
I'm totally on board with just supporting `Int`. I just wanted to ensure it wouldn't be too restrictive for some future use.

**Comment by chriselrod** [comment]
Just realized I also accidentally created a second branch called "static" where I added support for the integer types:
https://github.com/chriselrod/ArrayInterface.jl/tree/static
That one at chriselrod/ArrayInterface instead of at SciML/ArrayInterface like this PR.

**Comment by Tokazama** [comment]
> An advantage of the branchy style worth considering is that it is less prone to ambiguity errors.

It might be good for us to start using `detect_ambiguities` to weed these things out. If it doesn't affect performance and reduces ambiguities the branching style may be preferable, but there is something beautiful in the simplicity of all those one line methods.

**Comment by chriselrod** [comment]
I have noticed that using a lot of `@eval` loops can severely impact package load time, so I think it is normally preferable to keep the number of methods relatively small.

We have 36 definitions in the `+`/`-`/`*` loop.
The main savings we get is that, by not defining specific `::Static{0}` or `::Static{1}` definitions, we don't need to define any methods for the various combinations to avoid ambiguities.

I added
```julia
@test isempty(detect_ambiguities(ArrayInterface))
@test isempty(detect_unbound_args(ArrayInterface))
```

**Comment by timholy** [comment]
Can also use https://github.com/JuliaTesting/Aqua.jl

**Comment by Tokazama** [comment]
> Can also use https://github.com/JuliaTesting/Aqua.jl

That's really cool! I should be using that.

**Comment by chriselrod** [comment]
Sure, now using `Aqua.jl`.

**Comment by timholy** [comment]
@johnnychen94 introduced me to Aqua, I haven't yet used it.

@chriselrod, interesting. What does the `T >: Nothing` accomplish? Could you declare these
```julia
promote_rule(::Type{Nothing}, ::Type{S}) where S<:Static = ...
```

**Comment by chriselrod** [comment]
> What does the T >: Nothing accomplish?

Resolves a type ambiguity on master.

> Could you declare these

I already had definitions like
```julia
promote_rule(::Type{Nothing}, ::Type{<:Static}) = ...
```
Is there a difference in specificity?
I still have the same ambiguity errors.

**Comment by timholy** [comment]
Try rebuilding master? An ambiguity-related commit just got reverted within the last hour or so.

**Comment by chriselrod** [comment]
https://github.com/JuliaLang/julia/pull/37484

To what extant is it better to err on the side of conservative here?

> Anyway, the two former ones seem serious enough that is worth to revert this to get back to the "baseline" and then eventually put it back when the problems have been fixed.

Sounds like it would be better long term to be strict.

**Comment by chriselrod** [comment]
That said, no errors on Travis on 1.2, 1.5, or master now.
Presumably Travis hasn't updated master to reflect the reversion of the ambiguity commit yet (but I've tested locally before and after to confirm).

File: src/ranges.jl [hunk 1]
```diff
@@ -193,7 +203,7 @@ specified then indices for visiting each index of `x` is returned.
 """
 @inline function indices(x)
   inds = eachindex(x)
-  if inds isa AbstractUnitRange{<:Integer}
+  if inds isa AbstractUnitRange#{<:Integer} # prevents inference
```
Comment (timholy) [inline src/ranges.jl:206]: Could maybe use `inds isa AbstractUnitRange && eltype(inds) <: Integer`?

**Review by timholy — APPROVED** [review]
I didn't check in gory detail, but LGTM

File: src/ranges.jl [hunk 2]
```diff
@@ -44,6 +44,8 @@ known_step(::Type{<:AbstractUnitRange{T}}) where {T} = one(T)
 
 _get(x) = x
 _get(::Val{V}) where {V} = V
```
Comment (Tokazama) [inline src/ranges.jl:46]: Do we still need to support `Val` now that we're switching over to `Static` and `Val` can't be any of the fields in `OptionallyStaticUnitRange` anymore?

Comment (chriselrod) [reply src/ranges.jl]: I've removed it.

File: src/ranges.jl (hunk 1 above)
Comment (chriselrod) [reply src/ranges.jl]: Done, and inference tests pass.

**Comment by Tokazama** [comment]
I think this looks great.

**Pull request merged**